libs/vkd3d/command.c

```diff
@@ -18059,6 +18071,63 @@ static VkResult d3d12_command_queue_submit_split_locked(struct d3d12_device *dev
 	return VK_SUCCESS;
 }
 
+static void d3d12_command_queue_wait_staggered_submission(struct d3d12_command_queue *command_queue)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &command_queue->device->vk_procs;
+    VkSemaphoreWaitInfo wait_info;
+    VkResult vr;
+
+    if (!command_queue->last_submission_timeline_value)
+        return;
+
+    memset(&wait_info, 0, sizeof(wait_info));
+    wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO;
+    wait_info.semaphoreCount = 1;
+    wait_info.pSemaphores = &command_queue->vkd3d_queue->submission_timeline;
+    wait_info.pValues = &command_queue->last_submission_timeline_value;
+
+    if ((vr = VK_CALL(vkWaitSemaphores(command_queue->device->vk_device, &wait_info, UINT64_MAX))))
+        ERR("Failed to wait for semaphore, vr %d.\n", vr);
+
+    command_queue->last_submission_timeline_value = 0;
+}
+
+static bool d3d12_command_queue_needs_staggered_submissions_locked(struct d3d12_command_queue *command_queue)
+{
+    static const uint64_t inactive_threshold_ns = 1000000000ull; /* 1s */
+    uint64_t current_time_ns;
+
+    if (command_queue->vkd3d_queue->virtual_queue_count == 1)
```

Comment:
I wonder if a less risky implementation is to only consider a queue high priority if it has a swapchain associated with it. It's the only scenario I can think of where we actually want to go to this length.

Reply:
The current impl will also trigger for COMPUTE. I guess it will only trigger for cases where COMPUTE + COPY queues is more than 4. But then again, staggering work in async compute / copy is certainly less bad than direct queue.

Reply:
FWIW I added logging so we can easily see if we hit this path when we don't want to; given that queue priorities don't seem to be used much anyway this probably won't happen very often anyway.